samples/sample_common/include/sample_utils.h

```diff
@@ -119,110 +113,407 @@ bool IsDecodeCodecSupported(mfxU32 codecFormat);
+    {
+        if (!size || !id)
+            return nullptr;
 
-    CSmplYUVWriter();
-    virtual ~CSmplYUVWriter();
+        // Limitation: only one ExtBuffer instance can be stored
+        auto it = std::find_if(m_ext_buf.begin(), m_ext_buf.end(), CmpExtBufById(id));
+        if (it == m_ext_buf.end())
+        {
+            mfxExtBuffer* buf = (mfxExtBuffer*)new mfxU8[size];
+            memset(buf, 0, size);
+            m_ext_buf.push_back(buf);
 
-    virtual void      Close();
-    virtual mfxStatus Init(const msdk_char *strFileName, const mfxU32 numViews);
-    virtual mfxStatus Reset();
-    virtual mfxStatus WriteNextFrame(mfxFrameSurface1 *pSurface);
-    virtual mfxStatus WriteNextFrameI420(mfxFrameSurface1 *pSurface);
+            mfxExtBuffer& ext_buf = *buf;
+            ext_buf.BufferId = id;
+            ext_buf.BufferSz = size;
+            RefreshBuffers();
 
-    void SetMultiView() { m_bIsMultiView = true; }
+            return m_ext_buf.back();
+        }
 
-protected:
-    FILE         *m_fDest, **m_fDestMVC;
+        return *it;
+    }
+
+    mfxExtBuffer* GetExtBuffer(mfxU32 id) const
+    {
+        const auto it = std::find_if(m_ext_buf.begin(), m_ext_buf.end(), CmpExtBufById(id));
+        return it != m_ext_buf.end() ? *it : nullptr;
+    }
+
+    void RefreshBuffers()
+    {
+        this->NumExtParam = (mfxU32)m_ext_buf.size();
+        this->ExtParam    = this->NumExtParam ? m_ext_buf.data() : nullptr;
+    }
+
+    void ClearBuffers()
+    {
+        if (m_ext_buf.size())
+        {
+            for (auto it = m_ext_buf.begin(); it != m_ext_buf.end(); it++ )
+            {
+                delete [] (mfxU8*)(*it);
+            }
+            m_ext_buf.clear();
+        }
+        RefreshBuffers();
+    }
+
+    bool IsCopyAllowed(mfxU32 id)
+    {
+        static const mfxU32 allowed[] = {
+            MFX_EXTBUFF_CODING_OPTION,
+            MFX_EXTBUFF_CODING_OPTION2,
+            MFX_EXTBUFF_CODING_OPTION3,
+            MFX_EXTBUFF_FEI_PARAM,
+            MFX_EXTBUFF_BRC,
+            MFX_EXTBUFF_HEVC_PARAM,
+        };
+
+        auto it = std::find_if(std::begin(allowed), std::end(allowed),
+                               [&id](const mfxU32 allowed_id)
+                               {
+                                   return allowed_id == id;
+                               });
+        return it != std::end(allowed);
+    }
+
+    struct CmpExtBufById
+    {
+        mfxU32 id;
+
+        CmpExtBufById(mfxU32 _id)
+            : id(_id)
+        { };
+
+        bool operator () (mfxExtBuffer* b)
+        {
+            return  (b && b->BufferId == id);
+        };
+    };
+
+    std::vector<mfxExtBuffer*> m_ext_buf;
+};
+
+typedef ExtBufHolder<mfxVideoParam> MfxVideoParamsWrapper;
+typedef ExtBufHolder<mfxEncodeCtrl> mfxEncodeCtrlWrap;
+typedef ExtBufHolder<mfxInitParam>  mfxInitParamlWrap;
+typedef ExtBufHolder<mfxBitstream> _mfxBitstreamWrapper;
+
```

Comment:
As I see we don't use _mfxBitstreamWrapper anywhere but in these two lines. Maybe we can just squash them?

Reply:
Done